File: backend/schemas/bag.py
from pydantic import BaseModel, Field
from typing import Literal
from datetime import datetime


class BagItemCreate(BaseModel):
    product_id: int
    quantity: int = Field(..., gt=0)


class BagCreate(BaseModel):
    customer_name: str
    customer_phone: str | None = None
    items: list[BagItemCreate]
    note: str | None = None


class BagItemReturn(BaseModel):
    product_id: int
    quantity_sold: int = Field(0, ge=0)
    quantity_returned: int = Field(0, ge=0)


class BagReturn(BaseModel):
    items: list[BagItemReturn]
    payment: Literal["pix", "credito", "debito", "dinheiro"] = "pix"
    seller: str | None = None
    note: str | None = None


class BagItemResponse(BaseModel):
    id: int
    product_id: int
    product_name: str | None = None
    quantity_sent: int
    quantity_sold: int
    quantity_returned: int
    unit_price: float

    class Config:
        from_attributes = True


class BagResponse(BaseModel):
    id: int
    customer_name: str
    customer_phone: str | None = None
    status: str
    note: str | None = None
    payment: str | None = None
    seller: str | None = None
    total_sold_amount: float
    order_id: int | None = None
    date_out: datetime | None = None
    returned_at: datetime | None = None
    items: list[BagItemResponse]

    class Config:
        from_attributes = True

File: frontend/app/services/bag.ts
import { api } from "./api";

export interface BagItem {
  id: number;
  product_id: number;
  product_name?: string | null;
  quantity_sent: number;
  quantity_sold: number;
  quantity_returned: number;
  unit_price: number;
}

export interface Bag {
  id: number;
  customer_name: string;
  customer_phone?: string | null;
  status: string;
  note?: string | null;
  payment?: string | null;
  seller?: string | null;
  total_sold_amount: number;
  order_id?: number | null;
  date_out?: string | null;
  returned_at?: string | null;
  items: BagItem[];
}

export type BagCreatePayload = {
  customer_name: string;
  customer_phone?: string | null;
  note?: string | null;
  items: Array<{
    product_id: number;
    quantity: number;
  }>;
};

export type BagReturnPayload = {
  payment: "pix" | "credito" | "debito" | "dinheiro";
  seller?: string | null;
  note?: string | null;
  items: Array<{
    product_id: number;
    quantity_sold: number;
    quantity_returned: number;
  }>;
};

export async function getBags(): Promise<Bag[]> {
  const response = await api.get("/bags/");
  return response.data;
}

export async function getBag(id: number): Promise<Bag> {
  const response = await api.get(`/bags/${id}`);
  return response.data;
}

export async function createBag(payload: BagCreatePayload): Promise<Bag> {
  const response = await api.post("/bags/", payload);
  return response.data;
}

export async function returnBag(id: number, payload: BagReturnPayload): Promise<Bag> {
  const response = await api.post(`/bags/${id}/return`, payload);
  return response.data;
}